### PR TITLE
feat(zk): implement halo2 proof verification

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR implements halo2 proof verificaiton

### CODE REVIEW

File: tachyon/zk/plonk/permutation/permutation_verification.h [hunk 1]
```diff
@@ -48,8 +48,7 @@ size_t GetSizeOfPermutationVerificationExpressions(
     const ConstraintSystem<F>& constraint_system) {
   size_t num_products = constraint_system.ComputePermutationProductNums();
   if (num_products == 0) return 0;
-  size_t chunk_len = constraint_system.ComputePermutationChunkLen();
-  return num_products + 1 + chunk_len;
+  return 2 + 2 * num_products - 1;
```
Comment (Insun35) [inline tachyon/zk/plonk/permutation/permutation_verification.h:51]:
```suggestion
  return 2 * num_products + 1;
```

**Comment by TomTaehoonKim** [comment]
@chokobole Can you fix typo in [6608b07](https://github.com/kroma-network/tachyon/pull/228/commits/6608b0797c6e56e388ec6aa6a64689d9f6b87185)? eavl -> eval.

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/zk/plonk/halo2/verifier.h [hunk 2]
```diff
@@ -314,6 +316,71 @@ class Verifier : public VerifierBase<PCS> {
...
+
+  size_t GetSizeOfAdviceInstanceColumnQueries(
+      const ConstraintSystem<F>& constraint_system) {
+    const std::vector<AdviceQueryData>& advice_queries =
+        constraint_system.advice_queries();
+    const std::vector<InstanceQueryData>& instance_queries =
+        constraint_system.instance_queries();
+    size_t size = advice_queries.size();
+    if constexpr (PCS::kQueryInstance) {
+      size += instance_queries.size();
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/halo2/verifier.h:338]:
```suggestion
    size_t size = constraint_system.advice_queries().size();
    if constexpr (PCS::kQueryInstance) {
      size += constraint_system.instance_queries().size();
```

File: tachyon/zk/plonk/halo2/verifier.h [hunk 3]
```diff
@@ -375,6 +381,17 @@ class Verifier : public VerifierBase<PCS> {
...
+
+      std::vector<Opening> permutation_queries = CreatePermutationQueries<PCS>(
+          proof.ToPermutationVerificationData(i), constraint_system);
+      queries.insert(queries.end(),
+                     std::make_move_iterator(permutation_queries.begin()),
+                     std::make_move_iterator(permutation_queries.end()));
+    }
+
+    F expected_h_eval = ComputeExpectedHEval(num_circuits, vkey, proof);
+    if (expected_h_eval_out) {
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/zk/plonk/halo2/verifier.h:433]: To check the value from outside? in unittest?

File: tachyon/zk/plonk/permutation/permutation_verification.h [hunk 4]
```diff
@@ -104,6 +105,45 @@ std::vector<F> CreatePermutationVerificationExpressions(
...
+  for (size_t i = 0; i < num_products; ++i) {
+    queries.emplace_back(base::DeepRef<const C>(&data.product_commitments[i]),
+                         base::DeepRef<const F>(data.x), data.product_evals[i]);
+    queries.emplace_back(base::DeepRef<const C>(&data.product_commitments[i]),
+                         base::DeepRef<const F>(data.x_next),
+                         data.product_next_evals[i]);
+  }
+  if (num_products > 1) {
+    // Open it at ω^{last} * x for all but the last set.
+    for (size_t i = num_products - 2; i != SIZE_MAX; --i) {
```
[the remark this answers is not in the document]
Comment (TomTaehoonKim) [reply tachyon/zk/plonk/permutation/permutation_verification.h:139]: Oh I see you've added std_def.h for size_t! Can you also add stdint.h for `SIZE_MAX` as well?

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/zk/plonk/halo2/verifier.h [hunk 5]
```diff
@@ -314,6 +316,71 @@ class Verifier : public VerifierBase<PCS> {
...
+      openings.emplace_back(
+          base::DeepRef<const Commitment>(&commitments[column.index()]),
+          base::DeepRef<const F>(&points.back()), evals[i]);
+    }
+  }
+
+  bool DoVerify(
+      const std::vector<std::vector<Commitment>>& instance_commitments_vec,
+      const VerifyingKey<PCS>& vkey, const Proof<F, Commitment>& proof) {
+    std::vector<Opening> queries;
```
Comment (dongchangYoo) [inline tachyon/zk/plonk/halo2/verifier.h:364]: How about renaming 'queries' to 'openings'? That's why, we don't use `VerifierQuery` any more. 
(It is just recommendation, not necessary)

**Review by dongchangYoo — APPROVED** [review]
Else, LGTM

File: tachyon/zk/plonk/halo2/verifier.h (hunk 5 above)
Comment (chokobole) [reply tachyon/zk/plonk/halo2/verifier.h:364]: Yeap, I wanted so, too! But, at this moment, like `AdviceQuery`, we name all coupled with `XXXQuery`. So how about renaming all in the next iteration?

**Pull request merged**